pyro/distributions/hmm.py

```diff
@@ -130,3 +153,94 @@ def log_prob(self, value):
+    :param ~torch.distributions.MultivariateNormal transition_dist: A joint
+        distribution factor over a pair of successive time steps. This should
+        have batch_shape broadcastable to ``self.batch_shape + (num_steps,)``.
+        This should have event_shape ``(hidden_dim + hidden_dim,)`` (old+new).
+    :param ~torch.distributions.MultivariateNormal observation_dist: A joint
+        distribution factor over a hidden and an observed state. This should
+        have batch_shape broadcastable to ``self.batch_shape + (num_steps,)``.
+        This should have event_shape ``(hidden_dim + obs_dim,)``.
+    """
+    arg_constraints = {}
+
+    def __init__(self, initial_dist, transition_dist, observation_dist, validate_args=None):
+        assert isinstance(initial_dist, torch.distributions.MultivariateNormal)
+        assert isinstance(transition_dist, torch.distributions.MultivariateNormal)
+        assert isinstance(observation_dist, torch.distributions.MultivariateNormal)
+        hidden_dim = initial_dist.event_shape[0]
+        assert transition_dist.event_shape[0] == hidden_dim + hidden_dim
+        obs_dim = observation_dist.event_shape[0] - hidden_dim
+
+        shape = broadcast_shape(initial_dist.batch_shape + (1,),
+                                transition_dist.batch_shape,
+                                observation_dist.batch_shape)
+        batch_shape, time_shape = shape[:-1], shape[-1:]
+        event_shape = time_shape + (obs_dim,)
+        super(GaussianMRF, self).__init__(batch_shape, event_shape, validate_args=validate_args)
+        self.hidden_dim = hidden_dim
+        self.obs_dim = obs_dim
+        self._init = mvn_to_gaussian(initial_dist)
+        self._trans = mvn_to_gaussian(transition_dist)
+        self._obs = mvn_to_gaussian(observation_dist)
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(GaussianMRF, _instance)
+        batch_shape = torch.Size(broadcast_shape(self.batch_shape, batch_shape))
+        # We only need to expand one of the inputs, since batch_shape is determined
+        # by broadcasting all three. To save computation in _sequential_gaussian_tensordot(),
+        # we expand only _init, which is applied only after _sequential_gaussian_tensordot().
+        new._init = self._init.expand(batch_shape)
+        new._trans = self._trans
+        new._obs = self._obs
+        super(GaussianMRF, new).__init__(batch_shape, self.event_shape, validate_args=False)
+        new.validate_args = self.__dict__.get('_validate_args')
+        return new
+
+    def log_prob(self, value):
+        # Combine observation and transition factors.
+        result = self._trans
+        result += self._obs.condition(value).pad(left=self.hidden_dim)
+
+        # Eliminate time dimension.
+        result = _sequential_gaussian_tensordot(result)
+
+        # Combine initial factor.
+        result += self._init.pad(right=self.hidden_dim)
```

Comment:
nit: it'll be a bit faster if we combine using tensordot: `gaussian_tensordot(self._init, result, dims=self.hidden_dim)` and then we can marginalize out the final state. The reason is we'll only need to compute cholesky of 2 `hidden_dim x hidden_dim` matrices, instead of one `2hidden_dim x 2hidden_dim` matrix. This way also matches DiscreteHMM version.